src/util.c

```diff
@@ -2228,9 +2228,155 @@ nats_marshalULong(natsBuffer *buf, bool comma, const char *fieldName, uint64_t u
     return NATS_UPDATE_ERR_STACK(s);
 }
 
+// fmtFrac formats the fraction of v/10**prec (e.g., ".12345") into the
+// tail of buf, omitting trailing zeros. It omits the decimal
+// point too when the fraction is 0. It returns the index where the
+// output bytes begin and the value v/10**prec.
+static void
+fmt_frac(char *buf, int w, uint64_t v, int prec, int *nw, uint64_t *nv)
+{
+    // Omit trailing zeros up to and including decimal point.
+    bool print = false;
+    int i;
+    int digit;
 
-bool
-nats_IsSubjectValid(const char *subject, bool wcAllowed)
+    for (i = 0; i < prec; i++)
+    {
+        digit = v % 10;
+        print = print || digit != 0;
+        if (print)
+        {
+            w--;
+            buf[w] = digit + '0';
+        }
+        v /= 10;
+    }
+    if (print)
+    {
+        w--;
+        buf[w] = '.';
+    }
+    *nw = w;
+    *nv = v;
+}
+
+// fmtInt formats v into the tail of buf.
+// It returns the index where the output begins.
+static int
+fmt_int(char *buf, int w, uint64_t v)
+{
+    if (v == 0)
+    {
+        w--;
+        buf[w] = '0';
+    }
+    else
+    {
+        while (v > 0)
+        {
+            w--;
+            buf[w] = v % 10 + '0';
+            v /= 10;
+        }
+    }
+    return w;
+}
+
+natsStatus
+nats_marshalDuration(natsBuffer *out_buf, bool comma, const char *field_name, int64_t d)
+{
+    // Largest time is 2540400h10m10.000000000s
+    char buf[32];
+    int w = 32;
+    uint64_t u = d;
+    bool neg = d < 0;
+    int prec;
+    natsStatus s = NATS_OK;
+    const char *start = (comma ? ",\"" : "\"");
+
+    if (neg)
+        u = -u;
```

Comment:
We get a warning here on Windows since you define `u` as a uint:
```
warning C4146: unary minus operator applied to unsigned type, result still unsigned
```

Reply:
Still there.